MTM-64963 Rework MQTT Implementation for GA

scmi-c8y opened the pull request:
I'm not super-happy with this but I've been struggling to spend time on it, so I want to at least get it out there for people to start looking at. The reviewers are chosen because you have worked specifically on the MQTT side of the service, so you should have the knowledge to tell me what I've got wrong. I did find a few issues and questions while I was trying to verify specific behaviour but I will log those separately.

## Notes
1. The part you want to review is [here](https://cumulocity-iot-c8y-docs-build-pr-3925.surge.sh/device-integration/mqtt-service/#implementation)
2. There is a big red placeholder for details of the alarms that still needs to be completed. I welcome any input for this part.
3. The new `connecting-devices.md` file is also a placeholder that will be re-worked in the next PR. You don't need to review this.

scmi-c8y committed:
MTM-64963 Start reworking implementation section
* Move some text to new connecting devices section
* Re-do the first few implementation sub-sections
* Move remaining text into the new structure

scmi-c8y committed:
MTM-64963 Update a few quotas

scmi-c8y committed:
MTM-64963 MQTT v3.1.1 feature descriptions

scmi-c8y committed:
MTM-64963 Clarify that retained is not supported

scmi-c8y committed:
MTM-64963 Document support for MQTT 5 features
(still need to understand how Request Problem Information works)

scmi-c8y committed:
MTM-64963 Request Problem Information is supported

scmi-c8y committed:
MTM-64963 Topics and payloads

scmi-c8y committed:
MTM-64963 Clarify a limitation of SmartREST

scmi-c8y committed:
MTM-64963 Error reporting, quotas & limits

scmi-c8y committed:
MTM-64963 This section is still a placeholder

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
-| TLS    | 9883 |
-| no TLS | 2883 |
+MQTT devices can connect to the MQTT Service using direct TCP connections only.
+Authentication using "basic" (username/password) authentication and TLS client certificates is supported.
```
Authentication is duplicated:
```suggestion
Authentication using "basic" (username/password) credentials and TLS client certificates is supported.
```
or totally different wording

```suggestion
The service supports "basic" (username/password) and TLS client certificate authentication.
```

scmi-c8y replied:
I like the second suggestion, thanks!

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
+
+MQTT _retained messages_ are not supported by the MQTT Service.
+If the retain flag is set on a `PUBLISH` message from a device, the message will not be accepted and the connection will be closed.
+Devcies should not attempt to re-send an unacknowledged QoS 1 retained message after reconnecting, as this will simply cause the connection to be closed again.
```
```suggestion
Devices should not attempt to re-send an unacknowledged QoS 1 retained message after reconnecting, as this will simply cause the connection to be closed again.
```
On the other hand, I'm not sure if we need this sentence at all - it was initially misleading for me. Why the device may like to re-send retained messages if we don't accept it at all?

scmi-c8y replied:
Yeah, I think I will just delete this.

scmi-c8y replied:
Deleted in https://github.com/Cumulocity-IoT/c8y-docs/pull/3925/commits/53dc2c14cf41fa694e32fb10a0d0d19cbdb6b988

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
+#### Wildcard subscriptions {#wildcard-subscriptions}
+
+A wildcard subscription allows a device to subscribe to MQTT topics using a _pattern_ instead of a fixed topic name.
+The MQTT Service supports wildcard subscriptions using both the single-level (`+`) and multi-level (`#`) wildcard indicators.
```
Do we? :) I think that shouldn't be an issue, having a single topic, but until now, no one has really tested it.

scmi-c8y replied:
Well... we don't support it today but we've said that we will for the GA, so I guess this is a "hostage to fortune".

BeateRixen commented on content/device-integration/mqtt-service-bundle/connecting-devices.md:
```diff
@@ -0,0 +1,113 @@
+<br/>
+If the trust anchor (that is, the trusted root or intermediate certificate) used to validate the device certificate is trusted by multiple tenants, the device must also specify the tenant ID in the **MQTT username** field. This ensures that the platform can correctly identify which tenant the device is attempting to connect to. For more information, see [Authentication and authorization](#authent ...
+
+#### Adding and trusting CA certificate
```
```suggestion
#### Adding and trusting  CA (Certificate Authority) certificates {#ca-certificates}
```

scmi-c8y replied:
I guess everyone missed the line in the description saying you don't need to review this file. It will get re-worked by https://cumulocity.atlassian.net/browse/MTM-64961 that I will be starting on very soon. Anyway, I will apply all these suggestions to make the re-working easier.

BeateRixen commented on content/device-integration/mqtt-service-bundle/connecting-devices.md:
```diff
@@ -0,0 +1,113 @@
+<br/>
+Additionally, ensure that the **Auto registration** option is enabled when adding certificates. This allows any device presenting a valid certificate to be automatically registered on the platform when it first connects.
+
+#### Creating self-signed certificates
```
```suggestion
#### Creating self-signed certificates certificates {#creating-self-signed-certificates}
```

BeateRixen commented on content/device-integration/mqtt-service-bundle/connecting-devices.md:
```diff
@@ -0,0 +1,113 @@
+
+If you have more advanced requirements regarding certificate creation, see [Generating and signing certificates](/device-certificate-authentication/device-certificates#generating-and-signing-certificates).
+
+#### Using certificates
```
```suggestion
#### Using certificates certificates {#using-certificates}
```

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
+| Client Identifier            | Mandatory     | As for [version 3.1.1](#client-id).                                                                              |
+| Clean Start                  | Mandatory     | As for [version 3.1.1](#clean-session). _Clean Start_ is **required** on all device connections.                 |
+| Last Will                    | Supported     | With the same restrictions on QoS level, retained messages and device isolation as for [version 3.1.1](#last-will).<br>These additional version 5.0 properties on the will message are supported:<br>_Delay Interval_, _Payload Format Indicator_, _Content Type_, _Response Topic_, _Correlation Data_ and _User Properties_.<br>The _Message Expiry Interval ...
+| Receive Maximum              | Supported     | The MQTT Service will limit the number of unacknowledged QoS 1 messages for the device to the requested maximum. |
+| Maximum Packet Size          | Supported     | The MQTT Service will not send any message larger than the requested size to this device.<br>Note that messages larger than the reqeusted size will be **silently discarded**. |
+| Request Problem Information  | Supported     | A device should not assume that the MQTT Service will send a reason string, even when this has been requested.   |
```
We should double-check if the HiveMQ is supporting those out of the box as they were never tested.

scmi-c8y replied:
@schmiel-cumulocity I tested the maximum packet size - that works as expected. The other two are harder to test. I would strongly expect "receive maximum" to work (it would be a Hive bug if it didn't). As far as I understand it "request problem information" is somewhat optional anyway, so the server doesn't have to send a reason string even if the client requests it. We can probably claim it is supported even if the behavior doesn't change at all.

I'm pretty confident about the first two but could change "request problem information" to "Ignored" if you think that would be better?

schmiel-cumulocity replied:
You are right, let's leave it as it is.

BeateRixen commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
-MQTT Service requires clients to connect with clean session flag enabled, set to "1" (true), otherwise the client connection is rejected by the server.
-{{< /c8y-admon-important >}}
+The MQTT Service supports connections from clients using version 3.1, 3.1.1 or 5.0 of the MQTT protocol.
+Please refer to the [MQTT specifications](https://mqtt.org/mqtt-specification/) for details of the differences between these versions.
```
```suggestion
Refer to the [MQTT specifications](https://mqtt.org/mqtt-specification/) for details of the differences between these versions.
```

BeateRixen commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
-| no TLS | 2883 |
+MQTT devices can connect to the MQTT Service using direct TCP connections only.
+Authentication using "basic" (username/password) authentication and TLS client certificates is supported.
+For full details of the available ports and how to configure device authentication, see the [Connecting MQTT devices](/device-integration/mqtt-service/#connecting-devices) section.
```
```suggestion
For full details of the available ports and how to configure device authentication, see [Connecting MQTT devices](/device-integration/mqtt-service/#connecting-devices).
```

BeateRixen commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
+#### Quality of Service {#quality-of-service-qos}
 
-For subscriptions, the MQTT Service will deliver messages in the QoS that the client defined when subscribing to the topic (QoS 0 or 1).
+The MQTT Service supports two levels of MQTT _Quality of Service_ (QoS).
```
```suggestion
The MQTT Service supports two levels of MQTT Quality of Service (QoS).
```

BeateRixen commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
+
+| Feature                      | Support level | Notes                                                                                                            |
+|------------------------------|---------------|------------------------------------------------------------------------------------------------------------------|
+| Client Identifier            | Mandatory     | As for [version 3.1.1](#client-id).                                                                              |
```
```suggestion
| Client identifier            | Mandatory     | As for [version 3.1.1](#client-id).                                                                              |
```

BeateRixen replied:
@scmi-c8y Please check these tables for unnecessary capitalization. Most of the terms do not need capitalization, except for the first characters which we always capitalize. But I´m not 100% sure which ones are proper names.

scmi-c8y replied:
@BeateRixen Good question. I used the same capitalization as the MQTT specification, e.g. https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901046, which seems to treat the names of fields, packet types, etc. as proper names. I agree it doesn't necessarily align well with our style though. I'm happy to follow your guidance on this.

BeateRixen replied:
@scmi-c8y 
1. The spelling should definitely be consistent. The tables currently show both, for example, "Duplicate message indicator" and "Payload Format Indicator". 
2. I prefer small letters if possible, but if the MQTT spec treats these names as proper names, and we will always get confused about how to "transform" these into our styles, then let´s use capitalization as well, in all places where these terms are used.

scmi-c8y replied:
Resolved by merging #4047 (after some discussion about the right approach to use)

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
+| Correlation Data             | Supported     | Passed through between MQTT devices and Messaging Service clients.<br>Clients are responsible for including the correlation data on any response message(s). |
+| User properties              | Supported     | Passed through between MQTT devices and Messaging Service clients.                          |
+| Content Type                 | Supported     | Passed through between MQTT devices and Messaging Service clients.                          |
+| Message Expiry Interval      | Ignored       | The message expiry interval on messages published from devices will have no effect.<br>All messages published to devices will use the same message expiry interval. |
```
question: You mean Pulsar TTL here?

scmi-c8y replied:
Nope, this is separate to the Pulsar TTL.

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
+These topic name cannot be used by devices:
+
+* All _system topics_ (topic name beginning with `$`) unless specifically documented
+* `devicecontrol/notifications`
```
suggestion: In fact, it can be used, but it is only for SmartREST. In my opinion, we should move it to the `Core MQTT topics` chapter.

scmi-c8y replied:
Moved. What about `error`, is that allowed or not?

https://github.com/Cumulocity-IoT/c8y-docs/pull/3925/commits/53dc2c14cf41fa694e32fb10a0d0d19cbdb6b988

schmiel-cumulocity replied:
The error topic is not allowed, so it is fine. The full name is `$debug/$error`.

scmi-c8y replied:
Great, it's covered by the general restriction on system topics anyway so I removed it completely: https://github.com/Cumulocity-IoT/c8y-docs/pull/3925/commits/0ba1a0f344296f58aa83773a6b78048935bb3b1a. As you also mentioned, we are probably going to drop support for it anyway, and I think this part of the doc will be re-worked when we add the specific section for the SmartREST bridge, so I won't worry about it too much right now.

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
-Clients can connect using version 5.0 of the MQTT protocol.
-Support for additional MQTT 5.0 features will be added in future releases.
+There is no overlap between the Core MQTT and generic device topic spaces.
+However, to avoid potentially hard to diagnose failures, generic devices should avoid using any topic name starting with the Core MQTT prefixes `s/`, `t/`, `q/` or `c/`, even though some topics under those prefixes are not used by Core MQTT.
```
suggestion: We should probably also list JSON over MQTT topics here. They all start from `alarm/alarms/`, `event/events`, `measurement/measurements/`, `inventory/managedObjects/`

scmi-c8y replied:
Added, plus `devicecontrol/notifications`. Is that a complete list now? I might move this whole piece into the to-be-written section on Core MQTT later.

https://github.com/Cumulocity-IoT/c8y-docs/pull/3925/commits/53dc2c14cf41fa694e32fb10a0d0d19cbdb6b988

schmiel-cumulocity replied:
Yes, looks good 👍🏻

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
+See the [Service Quotas](/service-terms/quotas#mqtt-service) section for details of these limits.
+The soft quota can be increased up to the hard limit on request, although this may incur additional costs for the tenant.
+The message size used by these limits includes the message header as well as the payload.
+Message header size can vary significantly, particularly for MQTT version 5.0 devices, but it will always be at least 2 bytes, and usually more.
```
@abor-c8y should we add more details here? :)

abor-c8y replied:
I think this covers it. Although if we want to be more thorough I have found that the HiveMQ interceptors result in the following:

#### Extra Byte Allowance
##### MQTT5 Connect (Total of 25 bytes)
The following properties are resolved to mqtt spec defaults by the time interceptor receives the packet:
Variable header:
- 5 bytes for session expiry
- 3 bytes for receive maximum
- 5 bytes for maximum packet size
- 3 bytes for topic alias maximum
- 2 bytes for request response information
- 2 bytes for request problem information
Payload:
- 5 bytes for will delay interval

##### MQTT5 Publish (Total of 8 bytes)
The following properties are resolved to mqtt spec defaults by the time interceptor receives the packet:
Variable header:
- 3 bytes for topic alias
- 5 bytes for message expiry interval (what value do we get when we get this packet)

With the work that I have in review we will have to allow an extra 25 and 8 bytes for MQTT5 connect and publish packets respectively.

scmi-c8y replied:
Agreed, I think we can omit the details as we are being slightly generous. Nobody should complain that we are blocking packets that are actually below the limit.

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
+The MQTT Service follows the MQTT specification for responses from the server to devices.
 
-#### Adding and trusting CA certificate
+According to the specification, if the server receives a malformed packet or a protocol error, it must disconnect the device.
```
issue: That's true only for MQTT 3.1.1. MQTT 5.0 usually does not disconnect the client, it just sends back a negative response.

scmi-c8y replied:
I believe the _server_ is still required to disconnect on protocol violations. See https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901258, the highlighted bit.

However it was unclear how other rejected packets were handled so I have improved that. It duplicates some of the text from the "quotas and limits" section but I think that's OK, it's better that users read it twice rather than not at all.

https://github.com/Cumulocity-IoT/c8y-docs/pull/3925/commits/53dc2c14cf41fa694e32fb10a0d0d19cbdb6b988

abor-c8y commented on content/device-integration/mqtt-service-bundle/implementation.md:
```diff
@@ -4,274 +4,229 @@ title: MQTT protocol implementation
-Once the CA certificate has been uploaded and trusted in {{< product-c8y-iot >}}, devices can authenticate using client certificates signed by your trusted CA.
-To connect using any MQTT client, use the previously generated client certificate and key.
-This example uses the Mosquitto MQTT client:
+<font color="red" size="24">**TBC -- need details of the alarms**</font>
```
Currently alarms are raised when processing message from the `to-device` topic:

- `c8y_MqttService_ToDevice_NoKey` - Pulsar message has no key.
- `c8y_MqttService_ToDevice_InvalidKey` - Pulsar message has an unexpected key. For device isolation the key did not correspond to a subscribed clients id, for tenant isolation the message key did not match the topic.
- `c8y_MqttService_ToDevice_EmptyTopic` - Message property topic was an empty string.
- `c8y_MqttService_ToDevice_EmptyClientId` - Message property clientID was an empty string.

scmi-c8y replied:
Added to the Pulsar client doc in https://github.com/Cumulocity-IoT/c8y-docs/pull/3925/commits/4ef70d0817cc297319f2711a7f2b56238a0ea395

@abor-c8y What I was really looking for here was the `from-device` alarms. Do we know what those will be yet?

Also, wrt `c8y_MqttService_ToDevice_EmptyTopic`, does that also apply when the topic is _missing_ or just when it is _empty_? If the latter, what alarm do we raise when the topic is missing?

abor-c8y replied:
@scmi-c8y It seems I missed one, for missing topics we have `c8y_MqttService_ToDevice_MissingTopic`.

scmi-c8y replied:
Added in https://github.com/Cumulocity-IoT/c8y-docs/pull/3925/commits/207d524f8e33b74dba28d043f68282777ab40c7e

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Beate Rixen <90445236+BeateRixen@users.noreply.github.com>

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Beate Rixen <90445236+BeateRixen@users.noreply.github.com>
Co-authored-by: Sławomir Chmiel <slawomir.chmiel@cumulocity.com>

scmi-c8y committed:
MTM-64963 Add some to-device alarm descriptions
Not really part of the MQTT implementation but they came up during that
discussion.

scmi-c8y committed:
MTM-64963 Implement some tricky review comments

scmi-c8y committed:
MTM-64963 Add a missing to-device alarm type

schmiel-cumulocity reviewed (approved):
Good job! We can still change `error` topic name to the full name `$debug/$error`, but I'm fine either way, especially if we may drop support for it completely in the future.

scmi-c8y committed:
MTM-64963 No need to mention the error topic
explicitly as it is $debug/$error so covered by the general restriction
on system topics.

BeateRixen commented on content/device-integration/mqtt-service-bundle/connecting-devices.md:
```diff
@@ -0,0 +1,113 @@
+<br/>
+Additionally, ensure that the **Auto registration** option is enabled when adding certificates. This allows any device presenting a valid certificate to be automatically registered on the platform when it first connects.
+
+#### Creating self-signed certificates certificates {#creating-self-signed-certificates}
```
```suggestion
#### Creating self-signed certificates {#creating-self-signed-certificates}
```

BeateRixen commented on content/device-integration/mqtt-service-bundle/connecting-devices.md:
```diff
@@ -0,0 +1,113 @@
+
+If you have more advanced requirements regarding certificate creation, see [Generating and signing certificates](/device-certificate-authentication/device-certificates#generating-and-signing-certificates).
+
+#### Using certificates certificates {#using-certificates}
```
```suggestion
#### Using certificates {#using-certificates}
```

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Beate Rixen <90445236+BeateRixen@users.noreply.github.com>

scmi-c8y committed:
MTM-64963 MQTT keywords in CamelCase, MQTT-style. (#4047)

* MTM-64963 MQTT keywords in CamelCase, MQTT-style.
* MTM-64963 Use CamelCase for DUP flag too.
